### ISSUE TITLE
Major licensing issues

### ISSUE THREAD

**Issue by relistan** [newrelic/go-agent#45]
The license for the agent appears to indicate that you may only use the code in conjunction with your services:

> You may install, execute, and distribute these files and their contents only in conjunction with your direct use of New Relic’s services.

This makes sense for other agents but does not make sense for the Go agent, and appears overly restrictive.

You cannot simply remove the Go agent from an application, because Go code requires that you instrument the **code** all over the place by hand. This means the code as it stands can only be executed, even internally, as long as you have an valid subscription. It appears that you would be in violation of your license without removing all the calls to the agent code from your application in at least the following scenarios:

 * You instrumented an application with New Relic's Go agent and want to open source it. You can't run an instrumented application internally and provide it publicly as open source: they would have to be two different code bases or a patch set to add instrumentation. Also, for example, you cannot instrument any off the shelf AGPL-licensed application with the Go agent and run it in production.

 * You instrumented an application with the Go agent and your subscription lapses. Even if you can no longer benefit from New Relic's services, you would be in violation of the license without removing all the code from your application that calls the agent.

These are strong disincentives to using the Go agent. My belief from reading this license is that your intent is to prevent your own code from being used to compete with you either by helping a competitor or by supporting an open clone. That makes a lot of sense. Unfortunately the next line of the license is even more restrictive:

> These files and their contents shall not be used in conjunction with any other
product or software, including but not limited to those that may compete with
any New Relic product, feature, or software.

This is so broadly worded that it appears to actually prevent using the agent!

Can we get a legal review of this license? As it stands, as a long time, loyal New Relic customer we can't use the Go agent.

**Comment by relistan** [comment]
I should point out that these clauses are not present in at least two other agents:
 * the Ruby agent: https://github.com/newrelic/rpm/blob/master/LICENSE
 * the Node agent: https://github.com/newrelic/node-newrelic/blob/master/LICENSE

**Comment by victorNewRelic** [comment]
@relistan - thanks for the feedback on the license. I'm discussing this with our Legal team.

**Comment by relistan** [comment]
@victorNewRelic awesome, thanks!

**Comment by relistan** [comment]
Any update here @victorNewRelic?

**Comment by victorNewRelic** [comment]
Still working on it, @relistan. I'm targeting early next week to have an update for you.

**Comment by relistan** [comment]
Thanks @victorNewRelic

**Comment by victorNewRelic** [comment]
Hi @relistan,
We've talked this through internally and we feel the existing terms are still aligned with our requirements. The intent of those terms is to make sure the New Relic Go agent is used only with the New Relic services, whether paid or "Lite" subscriptions. In the case where your subscription lapses, the subscription reverts to "Lite" so you’re fine.
 
Your situation with regards to the license limiting what you can do with your open source project presents a different kind of use case. At the end of the day, what you need is for all your library dependencies to be compliant with your project's license. New Relic's license will not be broadly compatible with open source licenses because of the provision that the Go agent only be used in conjunction with New Relic services. I wonder though, would an open source project want to include vendor specific instrumentation?
 
I'm a huge fan of open source authors including instrumentation in their projects and there's a way that folks can do that with OpenTracing without tying their projects to specific vendor implementations or licenses. New Relic doesn't support OpenTracing yet though.
 
Thanks,
-Victor

**Comment by relistan** [comment]
Hi Victor. Thanks for reviewing but I think you guys missed the point of the complaint. The main thing is that this prevents anyone from using the agent at all:

>These files and their contents shall not be used in conjunction with any other
product or software, including but not limited to those that may compete with
any New Relic product, feature, or software.

The way this is worded, you cannot even use it  compiled with any another _library_. That doesn't make sense.

Secondly, in our case we need to run a piece of software that is licensed with the AGPL. This _requires_ that you open source any modifications to the software. Since we'd be instrumenting it with the Go agent, thus modifying it, we'd have to open source it. But we can't open source it under the terms of the New Relic agent license.

Third, in the case of general open source: it's not a problem at all to have New Relic instrumentation in a service as long as the instrumentation can be disabled. I'm not sure why you guys think that's a problem? Lots of things have instrumentation in them that people don't use.

**Comment by relistan** [comment]
@victorNewRelic Any response on this yet? I spoke with @benders who I believe was under the impression that there was some motion here. We're blocked on this, and honestly all of your customers should be blocked on this if they were to read the license.

**Comment by relistan** [comment]
@victorNewRelic if you guys aren't going to take action it would be good to know that for sure so we can give up on the Go agent. I'd rather get resolution instead, but either way an update would be great.

**Comment by nehaduggal** [comment]
@relistan I am sorry for the delay in responding while we continue to work on this. The terms you mentioned were put in the agreement to prevent misuse of our code. It was never our intention to confuse folks with our language.

Use of this agent has presented use cases that we had not foreseen and not considered when we initially drafted the license. We are working with our legal team on clarifying the language and getting you a license that will work for your use case.

**Comment by mihaitodor** [comment]
@nehaduggal @victorNewRelic Is there any update on this? It's been almost 6 months since this issue was raised and it would be good to know if we should expect to be able to use it in production anytime soon. Thank you!

**Comment by vinhnewrelic** [comment]
@mihaitodor @relistan I've been working on this with @nehaduggal . I understand the challenge folks are having with the language in the current license and had another conversation with our legal team. We were able to adjust the language so that it isn't as restrictive. This change has been merged into the build for the upcoming release (coming in the next few weeks).

Sorry about the wait! Let me know if you have any questions.
